src/sentry/digests/backends/redis.py

```diff
@@ -448,132 +198,51 @@ def maintenance(self, deadline, chunk=1000):
-                        make_last_processed_timestamp_key(timeline_key), self.ttl, int(time.time())
+        if timestamp is None:
+            timestamp = time.time()
+
+        connection = self._get_connection(key)
+        with self._get_timeline_lock(key, duration=30).acquire():
+            try:
+                response = script(
+                    connection, [key], ['DIGEST_OPEN', self.namespace, self.ttl, timestamp, key]
+                )
+            except ResponseError as e:
+                if 'err(invalid_state):' in e.message:
+                    six.raise_from(
+                        InvalidState('Timeline is not in the ready state.'),
+                        e,
                     )
-                    pipeline.execute()
-
-            def unschedule():
-                with connection.pipeline() as pipeline:
-                    # Watch the timeline to ensure that no other transactions add
-                    # events to the timeline while we are trying to delete it.
-                    pipeline.watch(timeline_key)
-                    pipeline.multi()
-                    if connection.zcard(timeline_key) == 0:
-                        cleanup_records(pipeline)
-                        pipeline.delete(make_last_processed_timestamp_key(timeline_key))
-                        pipeline.zrem(make_schedule_key(self.namespace, SCHEDULE_STATE_READY), key)
-                        pipeline.zrem(
-                            make_schedule_key(self.namespace, SCHEDULE_STATE_WAITING), key
-                        )
-                        pipeline.execute()
-
-            # If there were records in the digest, we need to schedule it so
-            # that we schedule any records that were added during digestion. If
-            # there were no items, we can try to remove the timeline from the
-            # digestion schedule.
-            if records:
-                reschedule()
-            else:
-                try:
-                    unschedule()
-                except WatchError:
-                    logger.debug('Could not remove timeline from schedule, rescheduling instead')
-                    reschedule()
-
-    def delete(self, key):
-        timeline_key = make_timeline_key(self.namespace, key)
-
-        connection = self.cluster.get_local_client_for_key(timeline_key)
-
-        lock = self.locks.get(timeline_key, duration=30, routing_key=timeline_key)
-        with lock.acquire(), connection.pipeline() as pipeline:
-            clear_timeline_contents(pipeline, timeline_key)
-            pipeline.zrem(make_schedule_key(self.namespace, SCHEDULE_STATE_READY), key)
-            pipeline.zrem(make_schedule_key(self.namespace, SCHEDULE_STATE_WAITING), key)
-            pipeline.execute()
+                else:
+                    raise
+
+            yield filter(
+                lambda record: record.value is not None,
+                map(
+                    lambda item: Record(
```

Comment:
Since it's a `namedtuple` how do you feel about giving the key names?

```py
Record(
    key=item[0],
    value=self.codec.decode(item[1]) if item[1] is not None else None,
    timestamp=float(item[2]),
)
```

The record generation before was a little better with the variable names being passed in as parameters:

https://github.com/getsentry/sentry/blob/12835ed8c3d8c1015622dadc304b3d9b4a46dc4c/src/sentry/digests/backends/redis.py#L514